src/Functions/generateSnowflakeID.cpp

```diff
@@ -0,0 +1,128 @@
+├─┼                 ┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┤
+|                   |     machine_id    |    machine_seq_num    |
+├─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┤
+
+- The first 41 (+ 1 top zero bit) bits is timestamp in Unix time milliseconds
+- The middle 10 bits are the machine ID.
+- The last 12 bits decode to number of ids processed by the machine at the given millisecond.
+*/
+
+constexpr auto timestamp_size = 41;
+constexpr auto machine_id_size = 10;
+constexpr auto machine_seq_num_size = 12;
+
+constexpr int64_t timestamp_mask = ((1LL << timestamp_size) - 1) << (machine_id_size + machine_seq_num_size);
+constexpr int64_t machine_id_mask = ((1LL << machine_id_size) - 1) << machine_seq_num_size;
+constexpr int64_t machine_seq_num_mask = (1LL << machine_seq_num_size) - 1;
+
+}
+
+class FunctionSnowflakeID : public IFunction
+{
+private:
+    mutable std::atomic<Int64> state{0};
+    // previous snowflake id
+    // state is 1 atomic value because we don't want use mutex
+
+public:
+    static constexpr auto name = "generateSnowflakeID";
+
+    static FunctionPtr create(ContextPtr /*context*/)
+    {
+        return std::make_shared<FunctionSnowflakeID>();
+    }
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 0; }
+
+    bool isDeterministicInScopeOfQuery() const override { return false; }
+    bool useDefaultImplementationForNulls() const override { return false; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+    bool isVariadic() const override { return true; }
+
+    bool isStateful() const override { return true; }
+    bool isDeterministic() const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        if (arguments.size() > 1) {
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Number of arguments for function {} doesn't match: passed {}, should be 0 or 1.",
+                getName(), arguments.size());
+        }
+
+        return std::make_shared<DataTypeInt64>();
+    }
+
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & /*arguments*/, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        auto col_res = ColumnVector<Int64>::create();
+        typename ColumnVector<Int64>::Container & vec_to = col_res->getData();
+        size_t size = input_rows_count;
+        vec_to.resize(size);
+
+        auto serverUUID = ServerUUID::get();
+
+        // hash serverUUID into 32 bytes
+        Int64 h = UUIDHelpers::getHighBytes(serverUUID);
+        Int64 l = UUIDHelpers::getLowBytes(serverUUID);
+        Int64 machine_id = ((h * 11) ^ (l * 17)) & machine_id_mask;
+
+        for (Int64 & el : vec_to) {
+            const auto tm_point = std::chrono::system_clock::now();
```

Comment:
This code will work too slow as it was in my first implementation of UUIDv7 generators because it calls clock_gettime behind the std::chrono facade for every SnowflakeID being generated. The better way would be getting a timestamp value only once for the given block and update it when the last_timestamp is greater than the current_timestamp.